src/fundamental_interface.jl

```diff
@@ -0,0 +1,275 @@
+###############################################################################
+#
+#   basic_interface.jl : basic interface for AbstractAlgebra
+#
+###############################################################################
+
+# TODO: Move more generic functions to this file.
+
+###############################################################################
+#
+#   Parents, elements and data type methods
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    parent(a)
+
+Return parent object of given element $a$.
+
+# Examples
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> G = SymmetricGroup(5); g = Perm([3,4,5,2,1])
+(1,3,5)(2,4)
+
+julia> parent(g) == G
+true
+
+julia> S, x = LaurentSeriesRing(ZZ, 3, "x")
+(Laurent series ring in x over Integers, x + O(x^4))
+
+julia> parent(x) == S
+true
+```
+"""
+function parent end
+
+# TODO: Give example
+@doc Markdown.doc"""
+    elem_type(parent)
+    elem_type(parent_type)
+
+Given a parent object (or its type), return the type of its elements.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> S, x = PowerSeriesRing(QQ, 2, "x")
+(Univariate power series ring in x over Rationals, x + O(x^3))
+
+julia> elem_type(S) == typeof(x)
+true
+```
+"""
+elem_type(x)  = elem_type(typeof(x))
+elem_type(T::DataType) = throw(MethodError(elem_type, (T,)))
+
+@doc Markdown.doc"""
+    parent_type(element)
+    parent_type(element_type)
+
+Given an element (or its type), return the type of its parent object.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> R, x = PolynomialRing(ZZ, "x")
+(Univariate Polynomial Ring in x over Integers, x)
+
+julia> S = MatrixSpace(R, 2, 2)
+Matrix Space of 2 rows and 2 columns over Univariate Polynomial Ring in x over Integers
+
+julia> a = rand(S, 0:1, 0:1);
+
+julia> parent_type(a) == typeof(S)
+true
+```
+"""
+parent_type(x) = parent_type(typeof(x))
+parent_type(T::DataType) = throw(MethodError(parent_type, (T,)))
+
+@doc Markdown.doc"""
+    base_ring(a::T)
```

Comment:
If `T` is not used in the definition and is not constrained, it seems better to leave it out (i.e. just write `base_ring(a)`).